File: embedded/pyexcel/kernel/framing.py
"""
Wire framing for the PyExcel v2 kernel protocol.

Frame layout (all integers little-endian, unsigned):

    +---------+----+----------+-----------+-----------+--------------+
    | body    | t  | meta_len | meta_json | n_payload | payloads...  |
    | u32     | u8 | u32      | bytes     | u32       |              |
    +---------+----+----------+-----------+-----------+--------------+

    body       = byte length of everything after this field (the frame body)
    t          = frame type, see FrameType
    meta_len   = byte length of meta_json
    meta_json  = UTF-8 encoded JSON object, application-defined per frame type
    n_payload  = number of binary payloads (typically 1 Arrow IPC stream per run)
    payload_i  = u32 size + raw bytes, repeated n_payload times

Design rules:

* **Pure stdlib.** No pyarrow/lxml imports here — those live in higher
  layers. Framing must stay loadable from the supervisor before any user
  dependency has been validated.
* **Bounded.** A maximum frame size is enforced on both encode and decode
  (default 256 MiB). A malformed peer cannot exhaust memory.
* **Stream-friendly.** ``read_frame`` reads exactly the bytes it needs from
  a ``read_exact`` callable. Partial reads are the caller's problem to
  compose with select/overlapped I/O.
* **Deterministic.** Same logical inputs always produce byte-identical
  frames; sorted JSON keys, no whitespace.

The C# counterpart in ``src/PyExcel.Bridge/Framing.cs`` (added in Phase 2)
must match this layout byte-for-byte. ``PROTOCOL_VERSION`` is the agreed
on-wire schema version both sides advertise in the ``Hello`` frame.
"""

from __future__ import annotations

import enum
import json
import struct
from dataclasses import dataclass, field
from typing import Callable, List, Tuple

PROTOCOL_VERSION = 2

DEFAULT_MAX_FRAME_BYTES = 256 * 1024 * 1024  # 256 MiB

_U8 = 1
_U32 = 4

_U32_LE = struct.Struct("<I")
_U8_LE = struct.Struct("<B")


class FrameType(enum.IntEnum):
    """Stable across the v2 protocol; new types append, existing never re-number."""

    HELLO = 1
    PING = 2
    PONG = 3
    SHUTDOWN = 4
    ERROR = 5

    RUN_REQUEST = 10
    RUN_RESULT = 11
    PROGRESS = 12
    LOG = 13
    CANCEL = 14
    LIST_JOBS = 15


class FramingError(Exception):
    """Base for all framing-layer errors."""


class FrameTooLargeError(FramingError):
    """Frame size exceeds the configured cap (caught *before* allocating)."""


class TruncatedFrameError(FramingError):
    """``read_exact`` returned fewer bytes than requested. Peer likely closed."""


class MalformedFrameError(FramingError):
    """Frame parsed but internal lengths are inconsistent."""


@dataclass(frozen=True)
class Frame:
    """One decoded wire frame.

    ``meta`` is always a parsed JSON object (dict).
    ``payloads`` is raw byte buffers in wire order.
    """

    type: FrameType
    meta: dict
    payloads: List[bytes] = field(default_factory=list)


# -----------------------------------------------------------------------------
# Encode
# -----------------------------------------------------------------------------


def encode_frame(
    frame_type: FrameType,
    meta: dict,
    payloads: Tuple[bytes, ...] = (),
    *,
    max_frame_bytes: int = DEFAULT_MAX_FRAME_BYTES,
) -> bytes:
    """Serialize a frame to a single ``bytes`` ready to write to the transport.

    Raises:
        FrameTooLargeError: encoded frame body would exceed ``max_frame_bytes``.
        TypeError: ``meta`` not JSON-serializable, or a payload isn't bytes-like.
        ValueError: ``frame_type`` not a member of :class:`FrameType`.
    """
    if not isinstance(frame_type, FrameType):
        try:
            frame_type = FrameType(frame_type)
        except ValueError as exc:
            raise ValueError(f"unknown frame type: {frame_type!r}") from exc

    meta_bytes = json.dumps(
        meta,
        ensure_ascii=False,
        separators=(",", ":"),
        sort_keys=True,
    ).encode("utf-8")

    for i, p in enumerate(payloads):
        if not isinstance(p, (bytes, bytearray, memoryview)):
            raise TypeError(f"payload {i} must be bytes-like, got {type(p).__name__}")

    payload_count = len(payloads)
    payload_bytes_total = sum(len(p) for p in payloads)

    body_len = (
        _U8
        + _U32
        + len(meta_bytes)
        + _U32
        + payload_count * _U32
        + payload_bytes_total
    )

    if body_len > max_frame_bytes:
        raise FrameTooLargeError(
            f"encoded frame body of {body_len} bytes exceeds cap of "
            f"{max_frame_bytes} bytes"
        )

    out = bytearray()
    out += _U32_LE.pack(body_len)
    out += _U8_LE.pack(int(frame_type))
    out += _U32_LE.pack(len(meta_bytes))
    out += meta_bytes
    out += _U32_LE.pack(payload_count)
    for p in payloads:
        out += _U32_LE.pack(len(p))
        out += bytes(p)

    return bytes(out)


# -----------------------------------------------------------------------------
# Decode
# -----------------------------------------------------------------------------


ReadExact = Callable[[int], bytes]
"""``read_exact(n)`` returns exactly ``n`` bytes, blocking as needed.

Returns shorter buffer only on clean EOF; framing treats short returns as
end-of-stream and raises :class:`TruncatedFrameError`.
"""


def read_frame(
    read_exact: ReadExact,
    *,
    max_frame_bytes: int = DEFAULT_MAX_FRAME_BYTES,
) -> Frame:
    """Read one frame from a stream.

    Raises:
        TruncatedFrameError: connection closed mid-frame.
        FrameTooLargeError: header announced a body larger than ``max_frame_bytes``.
        MalformedFrameError: header parsed but internal lengths inconsistent.
    """
    total_buf = read_exact(_U32)
    if len(total_buf) < _U32:
        raise TruncatedFrameError("EOF reading frame length prefix")
    (body_len,) = _U32_LE.unpack(total_buf)

    if body_len > max_frame_bytes:
        raise FrameTooLargeError(
            f"peer announced frame of {body_len} bytes; cap is {max_frame_bytes}"
        )
    if body_len < _U8 + _U32 + _U32:
        raise MalformedFrameError(f"frame body length {body_len} is below minimum")

    body = read_exact(body_len)
    if len(body) < body_len:
        raise TruncatedFrameError(
            f"EOF in frame body: expected {body_len} bytes, got {len(body)}"
        )

    offset = 0

    (type_byte,) = _U8_LE.unpack_from(body, offset)
    offset += _U8
    try:
        ftype = FrameType(type_byte)
    except ValueError as exc:
        raise MalformedFrameError(f"unknown frame type byte {type_byte}") from exc

    (meta_len,) = _U32_LE.unpack_from(body, offset)
    offset += _U32
    if offset + meta_len > body_len:
        raise MalformedFrameError(
            f"meta_len {meta_len} would read past end of frame body"
        )
    meta_bytes = bytes(body[offset : offset + meta_len])
    offset += meta_len

    try:
        meta = json.loads(meta_bytes.decode("utf-8")) if meta_len else {}
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise MalformedFrameError(f"meta is not valid UTF-8 JSON: {exc}") from exc

    if not isinstance(meta, dict):
        raise MalformedFrameError(
            f"meta must be a JSON object, got {type(meta).__name__}"
        )

    if offset + _U32 > body_len:
        raise MalformedFrameError("frame truncated before payload count")
    (n_payload,) = _U32_LE.unpack_from(body, offset)
    offset += _U32

    payloads: List[bytes] = []
    for i in range(n_payload):
        if offset + _U32 > body_len:
            raise MalformedFrameError(f"frame truncated before size of payload {i}")
        (psize,) = _U32_LE.unpack_from(body, offset)
        offset += _U32
        if offset + psize > body_len:
            raise MalformedFrameError(
                f"payload {i} size {psize} would read past end of frame body"
            )
        payloads.append(bytes(body[offset : offset + psize]))
        offset += psize

    if offset != body_len:
        raise MalformedFrameError(
            f"frame has {body_len - offset} trailing bytes after parsing"
        )

    return Frame(type=ftype, meta=meta, payloads=payloads)


def buffer_reader(buf: bytes) -> ReadExact:
    """Return a ``read_exact`` callable backed by an in-memory buffer.

    Used by tests and by the stdin-driven launcher mode. Yields ``b""``
    once the buffer is drained, which framing treats as EOF.
    """
    pos = [0]
    blen = len(buf)

    def _read(n: int) -> bytes:
        start = pos[0]
        end = min(start + n, blen)
        chunk = buf[start:end]
        pos[0] = end
        return chunk

    return _read

File: src/PyExcel.Bridge/Framing.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace PyExcel.Bridge;

/// <summary>
/// Wire framing for the PyExcel v2 kernel protocol.
///
/// Frame layout (all integers little-endian, unsigned):
/// <code>
///   +---------+----+----------+-----------+-----------+--------------+
///   | body    | t  | meta_len | meta_json | n_payload | payloads...  |
///   | u32     | u8 | u32      | bytes     | u32       |              |
///   +---------+----+----------+-----------+-----------+--------------+
/// </code>
/// Counterpart to <c>embedded/pyexcel/kernel/framing.py</c> — both sides
/// must encode and decode byte-for-byte identical frames.
///
/// Design rules (mirror the Python side):
/// <list type="bullet">
///   <item><b>Pure stdlib.</b> No third-party JSON dependency — see
///         <see cref="CanonicalJson"/>.</item>
///   <item><b>Bounded.</b> A configurable maximum frame size (default 256
///         MiB) is enforced on both encode and decode so a malformed peer
///         cannot exhaust memory.</item>
///   <item><b>Stream-friendly.</b> <see cref="ReadFrame"/> reads exactly the
///         bytes it needs from the supplied stream.</item>
///   <item><b>Deterministic.</b> Same logical inputs always produce
///         byte-identical frames; sorted JSON keys, no whitespace.</item>
/// </list>
/// </summary>
public static class Framing
{
    /// <summary>On-wire schema version advertised in the Hello frame.</summary>
    public const int ProtocolVersion = 2;

    /// <summary>Default cap on frame body length (256 MiB).</summary>
    public const int DefaultMaxFrameBytes = 256 * 1024 * 1024;

    private const int U8 = 1;
    private const int U32 = 4;

    // -------------------------------------------------------------------------
    // Encode
    // -------------------------------------------------------------------------

    /// <summary>
    /// Serialize a frame to a single byte array ready to write to the
    /// transport.
    /// </summary>
    /// <exception cref="FrameTooLargeException">
    /// Encoded frame body would exceed <paramref name="maxFrameBytes"/>.
    /// </exception>
    /// <exception cref="ArgumentException">
    /// <paramref name="frameType"/> is not a defined <see cref="FrameType"/>,
    /// or <paramref name="meta"/> contains a non-JSON-encodable value.
    /// </exception>
    public static byte[] EncodeFrame(
        FrameType frameType,
        IReadOnlyDictionary<string, object?> meta,
        IReadOnlyList<byte[]>? payloads = null,
        int maxFrameBytes = DefaultMaxFrameBytes)
    {
        if (meta is null) throw new ArgumentNullException(nameof(meta));
        if (!IsKnownFrameType(frameType))
            throw new ArgumentException($"unknown frame type: {(int)frameType}", nameof(frameType));

        payloads ??= Array.Empty<byte[]>();

        var metaBytes = CanonicalJson.Encode(meta);

        long payloadTotal = 0;
        for (var i = 0; i < payloads.Count; i++)
        {
            if (payloads[i] is null)
                throw new ArgumentException($"payload {i} is null", nameof(payloads));
            payloadTotal += payloads[i].Length;
        }

        long bodyLen = (long)U8
                     + U32
                     + metaBytes.Length
                     + U32
                     + (long)payloads.Count * U32
                     + payloadTotal;

        if (bodyLen > maxFrameBytes)
            throw new FrameTooLargeException(
                $"encoded frame body of {bodyLen} bytes exceeds cap of {maxFrameBytes} bytes");

        var totalLen = checked((int)(U32 + bodyLen));
        var output = new byte[totalLen];
        var offset = 0;

        WriteUInt32(output, ref offset, (uint)bodyLen);
        output[offset++] = (byte)frameType;
        WriteUInt32(output, ref offset, (uint)metaBytes.Length);
        Buffer.BlockCopy(metaBytes, 0, output, offset, metaBytes.Length);
        offset += metaBytes.Length;
        WriteUInt32(output, ref offset, (uint)payloads.Count);
        for (var i = 0; i < payloads.Count; i++)
        {
            var p = payloads[i];
            WriteUInt32(output, ref offset, (uint)p.Length);
            Buffer.BlockCopy(p, 0, output, offset, p.Length);
            offset += p.Length;
        }

        return output;
    }

    // -------------------------------------------------------------------------
    // Decode
    // -------------------------------------------------------------------------

    /// <summary>
    /// Read one frame from a stream. Blocks until the full frame is
    /// available or the stream returns EOF mid-frame.
    /// </summary>
    /// <exception cref="TruncatedFrameException">
    /// Stream returned EOF before the frame was complete.
    /// </exception>
    /// <exception cref="FrameTooLargeException">
    /// Peer announced a body larger than <paramref name="maxFrameBytes"/>.
    /// </exception>
    /// <exception cref="MalformedFrameException">
    /// Internal frame lengths are inconsistent or meta is not a JSON object.
    /// </exception>
    public static Frame ReadFrame(Stream input, int maxFrameBytes = DefaultMaxFrameBytes)
    {
        if (input is null) throw new ArgumentNullException(nameof(input));

        var lenBuf = ReadExact(input, U32, "frame length prefix");
        var bodyLen = ReadUInt32(lenBuf, 0);

        if (bodyLen > (uint)maxFrameBytes)
            throw new FrameTooLargeException(
                $"peer announced frame of {bodyLen} bytes; cap is {maxFrameBytes}");
        if (bodyLen < U8 + U32 + U32)
            throw new MalformedFrameException($"frame body length {bodyLen} is below minimum");

        var body = ReadExact(input, (int)bodyLen, "frame body");
        return DecodeBody(body);
    }

    /// <summary>
    /// Decode a frame from a complete in-memory wire buffer (length prefix
    /// included). Convenience for tests and round-trip code paths.
    /// </summary>
    public static Frame DecodeFrame(byte[] wire, int maxFrameBytes = DefaultMaxFrameBytes)
    {
        if (wire is null) throw new ArgumentNullException(nameof(wire));
        using var ms = new MemoryStream(wire, writable: false);
        return ReadFrame(ms, maxFrameBytes);
    }

    private static Frame DecodeBody(byte[] body)
    {
        var bodyLen = body.Length;
        var offset = 0;

        var typeByte = body[offset];
        offset += U8;
        var ftype = (FrameType)typeByte;
        if (!IsKnownFrameType(ftype))
            throw new MalformedFrameException($"unknown frame type byte {typeByte}");

        var metaLen = ReadUInt32(body, offset);
        offset += U32;
        if ((long)offset + metaLen > bodyLen)
            throw new MalformedFrameException(
                $"meta_len {metaLen} would read past end of frame body");

        Dictionary<string, object?> meta;
        if (metaLen == 0)
        {
            meta = new Dictionary<string, object?>(StringComparer.Ordinal);
        }
        else
        {
            var metaBytes = new byte[metaLen];
            Buffer.BlockCopy(body, offset, metaBytes, 0, (int)metaLen);
            offset += (int)metaLen;

            object? decoded;
            try
            {
                decoded = CanonicalJson.Decode(metaBytes);
            }
            catch (FormatException exc)
            {
                throw new MalformedFrameException(
                    $"meta is not valid UTF-8 JSON: {exc.Message}", exc);
            }
            if (decoded is not Dictionary<string, object?> obj)
                throw new MalformedFrameException(
                    $"meta must be a JSON object, got {decoded?.GetType().Name ?? "null"}");
            meta = obj;
        }

        if (offset + U32 > bodyLen)
            throw new MalformedFrameException("frame truncated before payload count");
        var nPayload = ReadUInt32(body, offset);
        offset += U32;

        var payloads = new List<byte[]>((int)nPayload);
        for (uint i = 0; i < nPayload; i++)
        {
            if (offset + U32 > bodyLen)
                throw new MalformedFrameException(
                    $"frame truncated before size of payload {i}");
            var psize = ReadUInt32(body, offset);
            offset += U32;
            if ((long)offset + psize > bodyLen)
                throw new MalformedFrameException(
                    $"payload {i} size {psize} would read past end of frame body");
            var p = new byte[psize];
            if (psize > 0)
                Buffer.BlockCopy(body, offset, p, 0, (int)psize);
            offset += (int)psize;
            payloads.Add(p);
        }

        if (offset != bodyLen)
            throw new MalformedFrameException(
                $"frame has {bodyLen - offset} trailing bytes after parsing");

        return new Frame(ftype, meta, payloads);
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private static byte[] ReadExact(Stream input, int n, string what)
    {
        var buf = new byte[n];
        var read = 0;
        while (read < n)
        {
            int r;
            try
            {
                r = input.Read(buf, read, n - read);
            }
            catch (IOException exc)
            {
                throw new TruncatedFrameException(
                    $"I/O failure reading {what}: {exc.Message}", exc);
            }
            if (r == 0)
                throw new TruncatedFrameException(
                    $"EOF reading {what}: expected {n} bytes, got {read}");
            read += r;
        }
        return buf;
    }

    private static void WriteUInt32(byte[] buf, ref int offset, uint value)
    {
        buf[offset] = (byte)value;
        buf[offset + 1] = (byte)(value >> 8);
        buf[offset + 2] = (byte)(value >> 16);
        buf[offset + 3] = (byte)(value >> 24);
        offset += 4;
    }

    private static uint ReadUInt32(byte[] buf, int offset)
    {
        return (uint)(buf[offset]
                    | buf[offset + 1] << 8
                    | buf[offset + 2] << 16
                    | buf[offset + 3] << 24);
    }

    private static bool IsKnownFrameType(FrameType t)
    {
        return t switch
        {
            FrameType.Hello
            or FrameType.Ping
            or FrameType.Pong
            or FrameType.Shutdown
            or FrameType.Error
            or FrameType.RunRequest
            or FrameType.RunResult
            or FrameType.Progress
            or FrameType.Log
            or FrameType.Cancel
            or FrameType.ListJobs => true,
            _ => false,
        };
    }
}
